src/Features/Core/Portable/PullMemberUp/QuickAction/AbstractMemberPullerWithQuickAction.cs

```diff
@@ -0,0 +1,187 @@
+            PullMembersUpAnalysisResult result,
+            ICodeGenerationService codeGenerationService,
+            SyntaxNode destinationNodeSyntax,
+            SolutionEditor solutionEditor,
+            Solution solution,
+            ImmutableDictionary<SyntaxTree, DocumentEditor> editorMap,
+            ImmutableDictionary<ISymbol, ImmutableArray<SyntaxNode>> syntaxMap)
+        {
+            // Add members to destination
+            var pullUpMembersSymbols = result.MembersAnalysisResults.SelectAsArray(memberResult => memberResult.Member);
+            var options = new CodeGenerationOptions(reuseSyntax: true, generateMethodBodies: false);
+            var membersAddedNode = codeGenerationService.AddMembers(destinationNodeSyntax, pullUpMembersSymbols, options: options);
+            var destinationEditor = editorMap[destinationNodeSyntax.SyntaxTree];
+            destinationEditor.ReplaceNode(destinationNodeSyntax, (syntaxNode, generator) => membersAddedNode);
+
+            // Remove the original members since we are pulling members into class
+            foreach (var analysisResult in result.MembersAnalysisResults)
+            {
+                foreach (var syntax in syntaxMap[analysisResult.Member])
+                {
+                    var originalMemberEditor = editorMap[syntax.SyntaxTree];
+                    originalMemberEditor.RemoveNode(originalMemberEditor.Generator.GetDeclaration(syntax));
+                }
+            }
+
+            return solutionEditor.GetChangedSolution();
+        }
+
+        private async Task<(ImmutableDictionary<SyntaxTree, DocumentEditor> editorMap, ImmutableDictionary<ISymbol, ImmutableArray<SyntaxNode>> syntaxMap)> InitializeEditorMapsAndSyntaxMapAsync(
+            PullMembersUpAnalysisResult result,
+            SyntaxNode destinationSyntaxNode,
+            SolutionEditor solutionEditor,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            // Members and destination may come from different documents,
+            // EditorMap is used to save and group all the editors will be used.
+            var editorMapBuilder = ImmutableDictionary.CreateBuilder<SyntaxTree, DocumentEditor>();
+            // One member may have multiple syntaxNodes (e.g partial method).
+            // SyntaxMap is used to find the syntaxNodes need to be changed more easily.
+            var syntaxMapBuilder = ImmutableDictionary.CreateBuilder<ISymbol, ImmutableArray<SyntaxNode>>();
+
+            foreach (var memberAnalysisResult in result.MembersAnalysisResults)
+            {
+                var tasks = memberAnalysisResult.Member.DeclaringSyntaxReferences.SelectAsArray(@ref => @ref.GetSyntaxAsync(cancellationToken));
+                var allSyntaxes = await Task.WhenAll(tasks).ConfigureAwait(false);
+                await AddEditorsToEditorMapBuilderAsync(editorMapBuilder, allSyntaxes.ToImmutableArray(), solutionEditor, solution);
```

Comment:
because this is all private, i wouldn't do the .ToImmutableArray here.  It doesn't really help.  You can just pass the single array you have without needing to make this copy.